runtime/lua/vim/treesitter/languagetree.lua

```diff
@@ -420,11 +468,31 @@ end
 ---     Set to `true` to run a complete parse of the source (Note: Can be slow!)
 ---     Set to `false|nil` to only parse regions with empty ranges (typically
 ---     only the root tree without injections).
---- @return table<integer, TSTree>
-function LanguageTree:parse(range)
+--- @param on_parse fun(trees?: table<integer, TSTree>, err?: string)? Function invoked when parsing completes.
+---     When provided, parsing will run asynchronously. The first argument to the function is the
+---     list of trees returned by the parse (upon success), or `nil` if the parse timed out
+---     (determined by 'redrawtime'). In the case of a failure, a string representing the error type
+---     will also be passed to the callback (currently only possible for timeouts).
+---
+---     If parsing was still able to finish synchronously (within 3ms), `parse()` returns the list
+---     of trees. Otherwise, it returns `nil`.
+--- @return table<integer, TSTree>?
```

Comment:
Just to note, I expect to change this in the future. I want to propose the convention that async-callback functions return an object with a close method to match UV. Detecting whether the parse happened without a schedule will need to be done some other way.